chctools/horn_hacking_lemmas.py

```diff
@@ -0,0 +1,710 @@
+import sys
+import z3
+import io
+
+import pysmt.environment
+import pysmt.solvers.z3 as pyz3
+from pysmt.smtlib.parser import SmtLibZ3Parser, Tokenizer
+
+def ground_quantifier(qexpr):
+    body = qexpr.body()
+
+    vars = list()
+    for i in reversed(range(qexpr.num_vars())):
+        vi_name = qexpr.var_name(i)
+        vi_sort = qexpr.var_sort(i)
+        vi = z3.Const(vi_name, vi_sort)
+        vars.append(vi)
+
+    body = z3.substitute_vars(body, *vars)
+    return (body, vars)
+
+def find_all_uninterp_consts(formula, res):
+    if z3.is_quantifier(formula):
+        formula = formula.body()
+
+    worklist = []
+    if z3.is_implies(formula):
+        worklist.append(formula.arg(1))
+        arg0 = formula.arg(0)
+        if z3.is_and(arg0):
+            worklist.extend(arg0.children())
+        else:
+            worklist.append(arg0)
+    else:
+        worklist.append(formula)
+
+    for t in worklist:
+        if z3.is_app(t) and t.decl().kind() == z3.Z3_OP_UNINTERPRETED:
+            res.append(t.decl())
+
+
+class HornRule(object):
```

Comment:
some of this code is mine, some is yours. Would be easier if separate code is in separate files. This way I get lost between reviewing my own code and what is new ;)

Reply:
Cool, I created a new pull request here: https://github.com/vhui/chc-tools/pull/2

Should be easier to compare everything I've changed against the base branch here.